File: ActionTest/ActionTest/MasterViewController.m
//
//  MasterViewController.m
//  ActionTest
//
//  Created by Nicolas Gomollon on 7/22/14.
//  Copyright (c) 2014 Techno-Magic. All rights reserved.
//

#import "MasterViewController.h"

@interface MasterViewController () {
	NSMutableArray *_objects;
}
@end

@implementation MasterViewController

- (void)loadView {
	[super loadView];
	self.title = @"Master";
	[self.tableView registerClass:[PTATableViewCell class] forCellReuseIdentifier:@"Cell"];
	_objects = [[NSMutableArray alloc] initWithArray:@[@"Swipe Me Left or Right", @"Swipe Me Left to Delete"]];
}

- (void)viewDidLoad {
	[super viewDidLoad];
	
	// Do any additional setup after loading the view, typically from a nib.
	self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(insertNewObject:)];
}

- (void)didReceiveMemoryWarning {
	[super didReceiveMemoryWarning];
	// Dispose of any resources that can be recreated.
}

- (void)insertNewObject:(id)sender {
	[_objects addObject:@"Swipe Me Left to Delete"];
	NSIndexPath *indexPath = [NSIndexPath indexPathForRow:(_objects.count - 1) inSection:0];
	[self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
}

- (UIView *)viewWithImageNamed:(NSString *)named {
	UIImageView *imageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:named]];
	imageView.contentMode = UIViewContentModeCenter;
	return imageView;
}

#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
	return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	return _objects.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	PTATableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
	
	cell.delegate = self;
	cell.textLabel.text = _objects[indexPath.row];
	
	if (indexPath.row == 0) {
		UIColor *greenColor = [UIColor colorWithRed:85.0f/255.0f green:213.0f/255.0f blue:80.0f/255.0f alpha:1.0f];
		
		[cell setPanGestureState:(PTATableViewCellStateLeftToRight | PTATableViewCellStateRightToLeft) mode:PTATableViewCellModeSwitch color:self.view.tintColor view:[self viewWithImageNamed:@"check"]];
		
		cell.leftToRightAttr.viewBehavior = PTATableViewCellSlidingViewBehaviorDragWithPanThenStick;
		cell.leftToRightAttr.color = greenColor;
		cell.rightToLeftAttr.rubberbandBounce = NO;
	} else {
		UIColor *redColor = [UIColor colorWithRed:232.0f/255.0f green:61.0f/255.0f blue:14.0f/255.0f alpha:1.0f];
		
		[cell setPanGestureState:PTATableViewCellStateLeftToRight mode:PTATableViewCellModeSwitch color:self.view.tintColor view:[self viewWithImageNamed:@"check"]];
		[cell setPanGestureState:PTATableViewCellStateRightToLeft mode:PTATableViewCellModeExit color:redColor view:[self viewWithImageNamed:@"cross"]];
		
		cell.rightToLeftAttr.triggerPercentage = 0.4f;
		cell.rightToLeftAttr.rubberbandBounce = NO;
		cell.rightToLeftAttr.viewBehavior = PTATableViewCellSlidingViewBehaviorDragWithPan;
	}
	
	return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
	[self.tableView deselectRowAtIndexPath:indexPath animated:YES];
	// Implement your own `tableView:didSelectRowAtIndexPath:` here.
}

#pragma mark - Pan Trigger Action (Required)

- (void)tableViewCell:(PTATableViewCell *)cell didTriggerState:(PTATableViewCellState)state withMode:(PTATableViewCellMode)mode {
	NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
	
	switch (mode) {
		case PTATableViewCellModeSwitch:
			NSLog(@"row %li's switch was triggered", (long)indexPath.row);
			break;
			
		case PTATableViewCellModeExit:
			NSLog(@"row %li's exit was triggered", (long)indexPath.row);
			[_objects removeObjectAtIndex:indexPath.row];
			[self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
			break;
			
		default:
			break;
	}
}

#pragma mark - Pan Trigger Action (Optional)

- (void)tableViewCellDidStartSwiping:(PTATableViewCell *)cell {
	NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
	NSLog(@"row %li started swiping", (long)indexPath.row);
}

- (void)tableViewCellIsSwiping:(PTATableViewCell *)cell withPercentage:(CGFloat)percentage {
	NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
	NSLog(@"row %li is being swiped with percentage: %0.1f", (long)indexPath.row, percentage * 100.0f);
}

- (void)tableViewCellDidEndSwiping:(PTATableViewCell *)cell {
	NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
	NSLog(@"row %li ended swiping", (long)indexPath.row);
}

@end

File: ActionTest/ActionTest/MasterViewController.swift
//
//  MasterViewController.swift
//  ActionTest
//
//  Created by Nicolas Gomollon on 6/18/14.
//  Copyright (c) 2014 Techno-Magic. All rights reserved.
//

import UIKit

class MasterViewController: UITableViewController, PTATableViewCellDelegate {
	
	var objects = ["Swipe Me Left or Right", "Swipe Me Left to Delete"]
	
	
	override func awakeFromNib() {
		super.awakeFromNib()
		tableView.register(PTATableViewCell.self, forCellReuseIdentifier: "Cell")
	}
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		// Do any additional setup after loading the view, typically from a nib.
		navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(MasterViewController.insertNewObject(_:)))
	}
	
	override func didReceiveMemoryWarning() {
		super.didReceiveMemoryWarning()
		// Dispose of any resources that can be recreated.
	}
	
	func insertNewObject(_ sender: AnyObject) {
		objects += ["Swipe Me Left to Delete"]
		let indexPath = IndexPath(row: (objects.count - 1), section: 0)
		tableView.insertRows(at: [indexPath], with: .automatic)
	}
	
	func viewWithImage(named: String) -> UIView {
		let imageView = UIImageView(image: UIImage(named: named))
		imageView.contentMode = .center
		return imageView
	}
	
	// MARK: - Table View
	
	override func numberOfSections(in tableView: UITableView) -> Int {
		return 1
	}
	
	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return objects.count
	}
	
	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath) as! PTATableViewCell
		
		cell.delegate = self
		cell.textLabel?.text = objects[indexPath.row]
		
		if indexPath.row == 0 {
			let greenColor = UIColor(red: 85.0/255.0, green: 213.0/255.0, blue: 80.0/255.0, alpha: 1.0)
			
			cell.setPanGesture([.leftToRight, .rightToLeft], mode: .switch, color: view.tintColor, view: viewWithImage(named: "check"))
			
			cell.leftToRightAttr.viewBehavior = .dragWithPanThenStick
			cell.leftToRightAttr.color = greenColor
			cell.rightToLeftAttr.rubberbandBounce = false
		} else {
			let redColor = UIColor(red: 232.0/255.0, green: 61.0/255.0, blue: 14.0/255.0, alpha: 1.0)
			
			cell.setPanGesture(.leftToRight, mode: .switch, color: view.tintColor, view: viewWithImage(named: "check"))
			cell.setPanGesture(.rightToLeft, mode: .exit, color: redColor, view: viewWithImage(named: "cross"))
			
			cell.rightToLeftAttr.triggerPercentage = 0.4
			cell.rightToLeftAttr.rubberbandBounce = false
			cell.rightToLeftAttr.viewBehavior = .dragWithPan
		}
		
		return cell
	}
	
	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		tableView.deselectRow(at: indexPath, animated: true)
		// Implement your own `tableView:didSelectRowAtIndexPath:` here.
	}
	
	// MARK: - Pan Trigger Action (Required)
	
	func tableView(cell: PTATableViewCell, didTrigger state: PTATableViewItemState, with mode: PTATableViewItemMode) {
		guard let indexPath = tableView.indexPath(for: cell) else { return }
		switch mode {
		case .switch:
			print("row \(indexPath.row)'s switch was triggered")
		case .exit:
			print("row \(indexPath.row)'s exit was triggered")
			objects.remove(at: indexPath.row)
			tableView.deleteRows(at: [indexPath], with: .fade)
		default:
			break
		}
	}
	
	// MARK: - Pan Trigger Action (Optional)
	
	func tableViewDidStartSwiping(cell: PTATableViewCell) {
		guard let indexPath = tableView.indexPath(for: cell) else { return }
		print("row \(indexPath.row) started swiping")
	}
	
	func tableViewIsSwiping(cell: PTATableViewCell, with percentage: Double) {
		guard let indexPath = tableView.indexPath(for: cell) else { return }
		print("row \(indexPath.row) is being swiped with percentage: \(percentage * 100.0)")
	}
	
	func tableViewDidEndSwiping(cell: PTATableViewCell) {
		guard let indexPath = tableView.indexPath(for: cell) else { return }
		print("row \(indexPath.row) ended swiping")
	}
	
}
